### ISSUE TITLE
Showing command instead of path in $0

### ISSUE THREAD

**Issue by tarungarg546** [yargs/yargs#538]
My environment details
- Windows 7
- Git Bash

Issue 

I am trying to make a CLI client for some use case. Now in that client i am using `yargs` like `below.

``` js
var argv = require('yargs')
  .usage(`Usage :-
  $0 -f [/path/to/keys/] -n [book-name] [author]`)
  .demand('f','Provide path to file that holds your keys.')
  .nargs('f',1)
  .describe('f','Path to file that holds your keys.')
  .demand('n','Provide a name you want to have.')
  .nargs('n',1)
  .describe('n','Name you want to have.')
  .demand(1,'Provide a author for which you want to')
  .help('h')
  .alias('h','help')
  .argv
```

Now for testing purpose i installed this globally using 

``` sh
npm ln
```

And when i am trying to run 

``` sh
myCommand -h
```

So now i would expect to have `myCommand` as a value for `$0` but actually i am getting something like 

``` sh
Usage :-
  Users\<username>\AppData\Roaming\npm\node_modules\<module>\index.js -f [/path/to/keys/] -n [book-name] [author]
```

looking into `yargs` code it looks like  `self.$0` is actually path to the file which is currently being executed.

But i want to have something like 

``` sh
Usage :-
  myCommand -f [/path/to/keys/] -n [book-name] [author]
```

this rather than path to file  , **is there any way to do it using yargs**?

And in this type of scenario should not `yargs` by default return `myCommand` by consulting the `bin` key from `package.json` from parent directory from which it is being invoked(I can take it up if this is a bug).

**Comment by cspotcode** [comment]
When commands are installed globally by npm, `process.argv` at runtime looks roughly like this:

```js
// On Unix
[ '/usr/local/bin/node', '/path/to/node/stuff/bin/cli-tool' ]
// On Windows
[ 'C:\Program Files\nodejs\node.exe', 'C:\Users\<username>\AppData\Roaming\npm\node_modules\<module name>\cli.js' ]
```

On Unix, argv[1] points directly to a file that's on our $PATH.  We can get the basename of argv[1] and check if `which cli-tool` resolves to the same path.  If so, we can substitute `cli-tool` for `$0`.  If not, we'll substitute the full absolute path.

On Windows, however, argv[1] is not the file on your %PATH%.  This is because NPM on Windows doesn't use symlinks.  It generates a .cmd (like a .bat) on your %PATH% that executes the script at argv[1].  I don't know of a reliable way to confirm at runtime whether or not our script is the same script on our %PATH% or not.

**Comment by cspotcode** [comment]
Here is the relevant yargs code: https://github.com/yargs/yargs/blob/master/yargs.js#L32-L47

**Comment by maxrimue** [comment]
I'm not fully sure about what to do here. Of course it is a problem that `$0` isn't very helpful when you use `npm ln` or `npm i -g .`. I think we have two options:
1. @cspotcode 's approach, verifying bin with `which`, or
2. Just grabbing the `package.json`'s `name`

While 1 requires some code which could introduce yet again more complexity and execution time, 2 is not _too_ reliable as `$0` would not necessarily be the actual globally available bin name. What do others think about this?

For the time being @tarungarg546, I guess you'll have to hardcode your app's name into the usage string :/

**Comment by GongT** [comment]
why not just give a option for *any* string as command name?

**Comment by cspotcode** [comment]
@GongT I don't think that's necessary.  If you want to explicitly specify the command name, don't use `$0`.  For example:

```
// Do this
var argv = require('yargs').usage('nameofcommand --options')
// Don't do this:
var argv = require('yargs').usage('$0 --options')
```

**Comment by josephfrazier** [comment]
I just ran into this issue as well, and ended up using [is-path-global](https://github.com/kevva/is-path-global) to substitute the basename when appropriate. Here's a minimal example:

```js
#!/usr/bin/env node

const isPathGlobal = require('is-path-global');
const yargs = require('yargs');

let scriptName = process.argv[1];
if (isPathGlobal(scriptName)) {
  scriptName = path.basename(scriptName);
} else {
  scriptName = '$0';
}

const argv = yargs
  .help()
  .usage(`Usage: ${scriptName} <command> [options]`)
  .argv;
```

**Comment by Igmat** [comment]
@maxrimue 

> For the time being @tarungarg546, I guess you'll have to hardcode your app's name into the usage string :/

It doesn't help in all cases because, this script name also is shown up in commands. Probably it's easier to add option for cli name?

**Comment by plroebuck** [comment]
@josephfrazier yeah, but that doesn't do much for non-local relative paths -- that said, I prefer just using `path.basename` regardless. Usage for a UN*X-based command only shows its basename, regardless of its invocation path.

Don't think **ANY** installation paths should be displayed via `--help` usage message.

**Comment by laggingreflex** [comment]
Fixed in #1143. You can now set it with `.scriptName()`
